### DISCUSSION TITLE
Quality enhancement of recorded audio using oboe

### DISCUSSION THREAD

**Discussion by ashfaq-opn** [google/oboe#1203] [General]
I am asking a very generic question. I compared the audio recorded with device's native audio recorder with input audio stream from oboe appended to a wav file. The audio recorded by oboe seemed low in quality than device's native recording app. I tried to tweak various stream settings, it didn't really help that much. As example my current settings are,

Sample Rate = 48000
Channel Count = oboe::ChannelCount::Stereo;
Audio API = oboe::AudioApi::AAudio;
Format = oboe::AudioFormat::I16;
Device Id = oboe::kUnspecified;
// 5ms of audio data at a sampling rate of 48kHz
Recording Frames Per Callback = 144

Can I get better quality (device recorder quality) audio recorded through oboe? Or this lower quality is a by-design feature for low latency audio stream?

Or this is because the native audio recorder does post-processing on the recorded audio? If it is the case, I can do the same in my app (the hard way), but I wanted to know if there is a way to get better quality by some settings only?

I want better quality because I want that people will use my app for karaoke singing or audio-story recording or poem recitation etc. I will definitely keep option to apply various digital filters, but of course the raw quality also matters.

**Comment by philburk** [comment]
Oboe should be able to achieve the same quality as any other API on Android. But there are many parameters than can affect quality.

> device's native audio recorder

What device are you using? A recorder provided by the manufacturer may be highly optimized.

> The audio recorded by oboe seemed low in quality

In what way? Did it have glitches and pops? Was it muffled? Was the volume too low? Was it clipped or distorted?
All of those have different causes.

> Format = oboe::AudioFormat::I16;

If you use AudioFormat::Float then you may be able to get 24-bit recording.

What InputPreset are you using? That can make a big difference because different presets apply different effects like Automatic Gain Control.  Try InputPreset:: VoicePerformance.

> // 5ms of audio data at a sampling rate of 48kHz
> Recording Frames Per Callback = 144

5 ms would be 240 frames. It would be better not to specify a framesPerCallback unless you absolutely require a specific size for your algorithm.

**Comment by ashfaq-opn** [comment]
@philburk I was in hurry to complete other issues in my app, Sorry for late response. Thank you so much. I will apply the preset and other things. I will reply with the improvements I observed.
